Increase strictness of global isel use for ROCM

tpopp opened the pull request:
Increase strictness while this is primarily for development instead of robust production cases (though it is still quite robust).

This will prevent fallback from global-isel with unsupported inputs to prevent a situation of attributing a witnessed change to improved global isel results that are actually caused by not using global isel or to prevent a lack of change caused by not actually utilizing global isel.

In this form, unsupported inputs will fail to compile entirely.

tpopp committed:
Increase strictness of global isel use for ROCM

Increase strictness while this is primarily for development instead of
robust production cases (though it is still quite robust).

This will prevent fallback from global-isel with unsupported inputs
to prevent a situation of attributing a witnessed change to improved
global isel results that are actually caused by not using global isel
or to prevent a lack of change caused by not actually utilizing global
isel.

In this form, unsupported inputs will fail to compile entirely.

Signed-off-by: Tres Popp <tpopp@brium.ai>

kuhar commented on compiler/plugins/target/ROCM/ROCMTarget.cpp:
```diff
@@ -462,7 +462,12 @@ class ROCMTargetBackend final : public TargetBackend {
         opt.NoInfsFPMath = false;
         opt.NoNaNsFPMath = true;
+        // Be extra cautious while this is less tested, and prevent unknown
+        // fallbacks from global isel.
```
Could you briefly explain what this option does and how we can observe the difference?

tpopp replied:
I've hopefully clarified the behavior with the elaborated comment. Please let me know what you think.

kuhar reviewed (commented):
I imagine this cannot be easily tested, right?

Looks good, just one suggestion.

tpopp committed:
Elaborate on the behavior of GlobalISelAbort

Signed-off-by: Tres Popp <tpopp@brium.ai>

tpopp committed:
Replace comment on FastISel with the actual fallback pass

Signed-off-by: Tres Popp <tpopp@brium.ai>

tpopp committed:
Add a test case for global isel abort behavior

This ensures that a failure that occurs in global isel does not fall
back to other options. This is tested by attempting to use an x86 intrinsic
which is unlikely for ROCM to support.

Signed-off-by: Tres Popp <tpopp@brium.ai>

tpopp committed:
Fixup test command to use global isel and test the failure

Signed-off-by: Tres Popp <tpopp@brium.ai>

kuhar reviewed (approved):
LGTM if tests pass
